include/bout/indexoffset.hxx

```diff
@@ -0,0 +1,50 @@
+#ifndef __INDEXOFFSET_H__
+#define __INDEXOFFSET_H__
+
+#include "bout/mesh.hxx"
+#include "bout/region.hxx"
+
+template <typename T = Ind3D>
+struct IndexOffset {
+
+  static_assert(std::is_base_of<Ind2D, T>::value || std::is_base_of<Ind3D, T>::value,
+                "IndexOffset must be templated with either Ind2D or Ind3D");
+
+  Mesh &mesh;
+  const int nx, ny, nz;
+
+  IndexOffset(Mesh &mesh)
+      : mesh(mesh), nx(mesh.LocalNx), ny(mesh.LocalNy), nz(mesh.LocalNz) {}
+
+  /// Convenience functions for converting to (x, y, z)
+  int x(T index) const { return (index.ind / nz) / ny; }
+  int y(T index) const { return (index.ind / nz) % ny; }
+  int z(T index) const { return (index.ind % nz); }
+
+  const inline T xp(T index, int i = 1) const { return index + (i * ny * nz); }
+  /// The index one point -1 in x
+  const inline T xm(T index, int i = 1) const { return index - (i * ny * nz); }
+  /// The index one point +1 in y
+  const inline T yp(T index, int i = 1) const { return index + (i * nz); }
+  /// The index one point -1 in y
+  const inline T ym(T index, int i = 1) const { return index - (i * nz); }
+  /// The index one point +1 in z. Wraps around zend to zstart
+  const inline T zp(T index, int i = 1) const {
+    ASSERT3(i > 0);
```

Comment:
Should we have these asserts for all directions or do we want to allow negative offsets?

Reply:
For x and y, I don't think it matters, as their implementations are symmetric, it's just for z that it seems to be important.

Reply:
It was more for consistency (i.e. easier to document).